sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -539,9 +541,10 @@ object ScalaReflection extends ScalaReflection {
 
         val params = getConstructorParameters(t)
         val fields = params.map { case (fieldName, fieldType) =>
-          if (javaKeywords.contains(fieldName)) {
-            throw new UnsupportedOperationException(s"`$fieldName` is a reserved keyword and " +
-              "cannot be used as field name\n" + walkedTypePath)
+          if (SourceVersion.isKeyword(fieldName) ||
```

Comment:
I'm not familiar with this method though, this check depends on users' Java runtime version? cc: @rednaxelafx @kiszk

Reply:
The `SourceVersion` class knows how to detect the spec version of the current running Java runtime, and can perform checks like identifier validation accordingly.
c.f. http://hg.openjdk.java.net/jdk8u/jdk8u/langtools/file/01036da3155c/src/share/classes/javax/lang/model/SourceVersion.java#l183

Reply:
Ah, I see. Thanks for the info, Kris. btw, the reserved keywords depend on Janino though, is it okay for the check to depend on running Java runtime?

Reply:
In the original implementation, `javaKeywords` contains `default` so how about checking keywords for Java 8 ?

Reply:
How about this, @cloud-fan ? I found he defined the initial set for the keywords in https://github.com/apache/spark/pull/13485

Reply:
While the reserved keywords depend on Janino though, `SourceVersion.isKeyword(fieldName)` would have the super set of keywords that Janino cannot accept. I think that this change is reasonable. We can avoid maintain `javaKeywords`.

Reply:
> SourceVersion.isKeyword(fieldName) would have the super set of keywords that Janino cannot accept

Looks nice, thanks for the check, @kiszk